### ISSUE TITLE
噪声池的梯度是否可以传回编码器用于训练，对水印模型性能影响大吗？

### ISSUE THREAD

**Issue by yokesky** [sh1newu/SepMark#11]
尊敬的大佬：
您好，您的论文思路非常新颖，让我大受启发。本人是一个研究深度伪造主动取证的新手，希望能请教一个问题。
```python
class Random_Noise(nn.Module):

    def __init__(self, layers, len_layers_R, len_layers_F):
        # 初始化函数，传入三个参数：layers，len_layers_R，len_layers_F
        super(Random_Noise, self).__init__()
        # 调用父类的初始化函数
        for i in range(len(layers)):
            # 遍历layers列表
            layers[i] = eval(layers[i])
            # 将layers列表中的元素转换为对应的类
        self.noise = nn.Sequential(*layers)
        # 将layers列表转换为Sequential模型
        self.len_layers_R = len_layers_R
        # 将len_layers_R赋值给self.len_layers_R
        self.len_layers_F = len_layers_F
        # 将len_layers_F赋值给self.len_layers_F
        print(self.noise)
        # 打印self.noise
        self.transform = transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])
        ])

    def forward(self, image_cover_mask): 
        # 将输入的image_cover_mask拆分为image、cover_image和mask
        image, cover_image, mask = image_cover_mask[0], image_cover_mask[1], image_cover_mask[2] 
        # 克隆并分离image、cover_image和mask
        forward_image = image.clone().detach() 
        forward_cover_image = cover_image.clone().detach() 
        forward_mask = mask.clone().detach() 
        # 创建三个全零的tensor，用于存储添加噪声后的图像 
        noised_image_C = torch.zeros_like(forward_image) 
        noised_image_R = torch.zeros_like(forward_image) 
        noised_image_F = torch.zeros_like(forward_image) 

        # 遍历每个图像 
        for index in range(forward_image.shape[0]):
            # 随机选择一个噪声层
            random_noise_layer_C = np.random.choice(self.noise, 1)[0]
            random_noise_layer_R = np.random.choice(self.noise[0:self.len_layers_R], 1)[0]
            random_noise_layer_F = np.random.choice(self.noise[self.len_layers_R:self.len_layers_R + self.len_layers_F], 1)[0]
            # 对每个图像添加噪声
            noised_image_C[index] = random_noise_layer_C([forward_image[index].clone().unsqueeze(0), forward_cover_image[index].clone().unsqueeze(0), forward_mask[index].clone().unsqueeze(0)])
            noised_image_R[index] = random_noise_layer_R([forward_image[index].clone().unsqueeze(0), forward_cover_image[index].clone().unsqueeze(0), forward_mask[index].clone().unsqueeze(0)])
            noised_image_F[index] = random_noise_layer_F([forward_image[index].clone().unsqueeze(0), forward_cover_image[index].clone().unsqueeze(0), forward_mask[index].clone().unsqueeze(0)])

            '''single_image = ((noised_image_C[index].clamp(-1, 1).permute(1, 2, 0) + 1) / 2 * 255).add(0.5).clamp(0, 255).to('cpu', torch.uint8).numpy()
            im = Image.fromarray(single_image)
            read = np.array(im, dtype=np.uint8)
            noised_image_C[index] = self.transform(read).unsqueeze(0).to(image.device)

            single_image = ((noised_image_R[index].clamp(-1, 1).permute(1, 2, 0) + 1) / 2 * 255).add(0.5).clamp(0, 255).to('cpu', torch.uint8).numpy()
            im = Image.fromarray(single_image)
            read = np.array(im, dtype=np.uint8)
            noised_image_R[index] = self.transform(read).unsqueeze(0).to(image.device)

            single_image = ((noised_image_F[index].clamp(-1, 1).permute(1, 2, 0) + 1) / 2 * 255).add(0.5).clamp(0, 255).to('cpu', torch.uint8).numpy()
            im = Image.fromarray(single_image)
            read = np.array(im, dtype=np.uint8)
            noised_image_F[index] = self.transform(read).unsqueeze(0).to(image.device)

        noised_image_gap_C = noised_image_C - forward_image
        noised_image_gap_R = noised_image_R - forward_image
        noised_image_gap_F = noised_image_F - forward_image'''
        # 对添加噪声后的图像进行裁剪，使其范围在-1到1之间
        noised_image_gap_C = noised_image_C.clamp(-1, 1) - forward_image
        noised_image_gap_R = noised_image_R.clamp(-1, 1) - forward_image
        noised_image_gap_F = noised_image_F.clamp(-1, 1) - forward_image

        # 返回添加噪声后的图像
        return image + noised_image_gap_C, image + noised_image_gap_R, image + noised_image_gap_F
```
\
我注意到您在这段代码中，加噪是通过**image + noised_image_gap** 实现的，其中的 noised_image_gap 是通过 **forward_image = image.clone().detach()** 计算得到的，这里的forward_image经过了梯度分离操作，也就是说实际上 反向传播时，解码器通过 image张量 向编码器传回梯度，而中间噪声池实际上没有向编码器传回梯度。
请问这样对于水印模型的性能影响大吗？因为我看了其他几篇论文（HiDDeN、Stetastamp），他们的噪声池梯度都是可以回传到编码器，也就是说编码器可以识别噪声池的攻击模式而主动编码出对应的鲁棒水印。您的这种设计方式似乎是将所有攻击模式置于一个黑盒中，没有白盒攻击。也就是说鲁棒性几乎完全是由解码器决定的。
但是似乎您的水印模型性能并没有比其他的噪声池可以回传梯度的模型出现性能下降，请问这是为什么？
请问我的理解对吗？您对于这个问题有什么看法？
希望能得到您的答复，感激不尽。

**Comment by sh1newu** [comment]
Thanks for your question. This trick is largely borrowed from the Straight-Through Estimator (also see Forward ASL, https://doi.org/10.1145/3474085.3475628). It is all right that during backpropagation, the gradient from the decoder is propagated directly to the encoder without passing through the noise layer. However, this allows the encoder and decoder to be trained jointly in an end-to-end manner, while the noise layer remains involved in the forward propagation. Therefore, I do not fully agree that the robustness is almost determined by the decoder. In my humble opinion, the encoder, the noise layer, the decoder, and even the discriminator all contribute to the robust training of the watermarking framework, and in essence, there is usually a trade-off between robustness and visual quality. Forward ASL has conducted primary experiments and concluded that when the attack distortion is known and fixed, forward ASL can be slightly inferior to the standard ASL but forward ASL is still a competitive approach. Nevertheless, without a thorough evaluation, it is not easy to assert which implementation is definitely better than the others, especially given the wide variety of network structures, losses, and hyperparameters, etc. In all, the implementation here is simple, straightforward, and effective for our task. We do not rule out and look forward to better schemes in this regard!

**Comment by yokesky** [comment]
Thank you for your response, it has been very beneficial to me.
